docs/CADL.md

```diff
@@ -0,0 +1,93 @@
+
+To represent cadl.json in memory, generator deserializes JSON into InputModel object graph. Then, with respect of configuration settings and customizations extracted from partial types and assembly level attributes (via `SourceInputModel`), generator creates OutputModel object graph. Various entities from OutputModel will be used by code writers to generate C# files. 
+
+InputModel is a slim, unvalidated representation of the cadl.json in memory. OutputModel is a rich, self-consistent set of objects that has all the data code writers need to generate valid C# code.
+
+With the exception of enums, no object should belong to InputModel and OutputModel at the same time.
+
+#### InputModel requirements
+- Minimalistic design. Only nessesary data is available, data duplication should be avoided (e.g. `InputModelType.DiscriminatorPropertyName` makes `InputModelProperty.IsDiscriminator` unnessesary). 
+- All InputModel objects are init-immutable (instance can't be modified after becoming accessible to external code).
+- InputModel is final. All decorator-based and configuration-based transformations and rules must be applied in emitter.
+- Lazy initializations must be avoided.
+
+#### OutputModel requirements
+- Rich design. Same data can be accessable and represented in multiple ways, depending on the needs of code writers (e.g. `ModelTypeProvider` and `JsonObjectSerialization` represent the same data for different code writers).
+- OutputModel parts can be created conditionaly and on demand. However, on-demand instantiation should be explicit, lazy initialization must be avoided (e.g. `LowLevelClient` has `CreateSerialization()` method rather than `Serialization` property).
+- Sufficient. OutputModel objects should have no properties or methods that return InputModel objects.
+- Decoupled from code writers. OutputModel object should not assume (explicitly or implicitly) what kind of writer is going to use it (e.g. protocol methods can be generated for both DPG clients and HLC Rest clients).
+
+## Unification of CADL and Swagger pipelines
+
+In Swagger pipeline, generator input is defined by [CodeModel specification](https://github.com/Azure/autorest/blob/main/packages/libs/codemodel/.resources/all-in-one/json/code-model.json). CodeModel.yaml itself is written in YAML. CodeModel types are over-complicated yet they aren't good enough to represent some schemas that can be defined in CADL. To share generation logic and types between Swagger pipeline and CADL pipeline, CodeModel object graph is transformed into InputModel object graph by `CodeModelConverter`. 
+
+#### Unification requirements
+
+- InputModel is sufficient. Its objects should have no properties or methods that return CodeModel objects. CodeModel-only parts of InputModel must have meaningful default values for CADL pipeline.  
+- All CodeModel rules and transformations must be applied before `CodeModelConverter` is invoked.
+
+## CADL + Swagger pipeline diagram
```

Comment:
what is the process line `sidecar->codewriter->codegen target`? Is it the process to generate `configuration.json`?

Reply:
Good point, I'll add `configuration.json` as a block

Reply:
Fixed